rtrlib/aspa/aspa.c

```diff
@@ -0,0 +1,267 @@
+/*
+ * This file is part of RTRlib.
+ *
+ * This file is subject to the terms and conditions of the MIT license.
+ * See the file LICENSE in the top level directory for more details.
+ *
+ * Website: http://rtrlib.realmv6.org/
+ */
+
+#include "aspa_private.h"
+#include "ordered_dyn_array/ordered_dyn_array.h"
+#include "rtrlib/rtrlib_export_private.h"
+#include "rtrlib/lib/alloc_utils_private.h"
+
+#include <pthread.h>
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+
+static void aspa_table_notify_clients(struct aspa_table *aspa_table, const struct aspa_record *record, const struct rtr_socket *rtr_socket, const bool added)
+{
+	if (aspa_table->update_fp) {
+		// Realloc in order not to expose internal record
+		struct aspa_record *rec = lrtr_realloc((void *)record, sizeof(record) + record->provider_count * sizeof(record->provider_asns[0]));
+		aspa_table->update_fp(aspa_table, *rec, rtr_socket, added);
+	}
+}
+
+RTRLIB_EXPORT void aspa_table_init(struct aspa_table *aspa_table, aspa_update_fp update_fp)
+{
+	aspa_table->update_fp = update_fp;
+	pthread_rwlock_init(&(aspa_table->lock), NULL);
+}
+
+int aspa_table_cache_init(struct aspa_table *aspa_table, struct rtr_socket *rtr_socket)
```

Comment:
i don't see why aspa_table should act as both a cache and a "fake object". as the cache doesn't use the update function, they're for different purposes and a pointless union type. i see that it might be cool to have the same add and remove function for both, yet i find this to be rather confusing as it's not immediately clear to the caller where the data is written to / deleted from.

Reply:
the cache however must be the same type as as the table, have a look at `rtr_sync_receive_and_store_pdus`

Reply:
the entire thing of writing records into sockets rather than the `aspa_table` quite a huge source of weirdness all over

Reply:
to make things clear -- our approach is inherently flawed. The only way to resolve this would be to (logical and)

1. make `rtr_socket` not store an `ordered_dyn_array`
2. create a new data structure that maps each socket address, that is, `struct rtr_socket *`, to an  `ordered_dyn_array` 
3. make aspa_table hold a reference to that data structure and (ofc) make aspa_table maintain that data structure.

- Note: In doing so we'd need to **iterate over this data structure to find** the correct `ordered_dyn_array` to add to order remove records from. Only this way we could avoid having all of this half-baked hacky cache-related weirdness all over the place.

Reply:
I really don't get the problem how about we just: (logical and)

1.) the `aspa_table` will own all `ordered_dyn-arrays` and partially manage them 
2.) rtr_sockets fetch one `ordered_dyn_array` from the aspa_table on their creation
3.) make `ordered_dyn_array` thread save
4.) when a rtr interaction happens, cache results -> lock the array -> add / remove all of them 

(we want most probably a Read-Write Lock) because of performance reasons.

Reply:
to your discussion: i mostly agree with @carl-tud that it's better to have as much ownership by the sockets and as little indirections/extra computation as possible.

> the cache however must be the same type as as the table, have a look at rtr_sync_receive_and_store_pdus

I don't quite agree this is the best way: it is not necessary to have this cache logic, if we don't stick with the pattern too strictly. we could use an ordered_dyn_array as both the main and shadow store in rtr_sync_receive_and_store_pdus. This would require some different function headers but would prevent the checking of an aspa table is a cache or not.

please, correct me if I'm wrong / am missing something :)

Reply:
I suggest: we stick with the dual-disponse mechanism for this pr, but do it on the ordered_dyn_array level, not on the aspa_table level. this would require different function headers but would prevent the strange intransparent dual use of aspa_table (with and without storing functionality)